web-prover-core/src/parser/extractors/types.rs

```diff
@@ -0,0 +1,195 @@
+//! # Extractor Types
+//!
+//! This module defines the common types used by extractors.
+
+use std::{collections::HashMap, fmt::Display};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tracing::debug;
+
+use crate::parser::{
+  extractors::get_value_type, predicate, predicate::Predicate, DataFormat, ExtractorConfig,
+  ExtractorError,
+};
+
+/// Trait for extracting data from a document
+pub trait DocumentExtractor {
+  fn validate_input(&self, data: &[u8]) -> Result<(), ExtractorError>;
+  fn extract(
+    &self,
+    data: &[u8],
+    config: &ExtractorConfig,
+  ) -> Result<ExtractionResult, ExtractorError>;
+}
+
+/// The type of data being extracted
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum ExtractorType {
+  /// String type
+  String,
+  /// Number type
+  Number,
+  /// Boolean type
+  Boolean,
+  /// Array type
+  Array,
+  /// Object type
+  Object,
+}
+
+impl ExtractorType {
+  pub fn is_valid_type(&self, value: &Value) -> Result<(), ExtractorError> {
+    let actual_type = get_value_type(value);
+    let expected_type_str = self.to_string();
+
+    if actual_type != expected_type_str {
+      return Err(ExtractorError::TypeMismatch {
+        expected: expected_type_str,
+        actual:   actual_type.to_string(),
+      });
+    }
+
+    Ok(())
+  }
+}
+
+impl TryFrom<&str> for ExtractorType {
+  type Error = ExtractorError;
+
+  fn try_from(value: &str) -> Result<Self, Self::Error> {
+    match value {
+      "string" => Ok(ExtractorType::String),
+      "number" => Ok(ExtractorType::Number),
+      "boolean" => Ok(ExtractorType::Boolean),
+      "array" => Ok(ExtractorType::Array),
+      "object" => Ok(ExtractorType::Object),
+      _ => Err(ExtractorError::UnsupportedExtractorType(value.to_string())),
+    }
+  }
+}
+
+impl Display for ExtractorType {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    let str = match self {
+      ExtractorType::String => "string",
+      ExtractorType::Number => "number",
+      ExtractorType::Boolean => "boolean",
+      ExtractorType::Array => "array",
+      ExtractorType::Object => "object",
+    }
+    .to_string();
+    write!(f, "{}", str)
+  }
+}
+
+/// Function to provide the default value `true` for `required`
+fn default_true() -> bool { true }
+
+/// A data extractor configuration
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Extractor {
```

Comment:
It almost seems like having `Extractor` be generic over a `Parser` or something would be handy in the future? Especially if we have nested extraction to do. i.e., JSON inside of HTML. 

Don't need to do it now, I'm just pontificating.

Reply:
I agree. I think we need more time user-testing, etc. before we introduce more changes/features.